File: src/fi/tuska/jalkametri/gui/DrinkDetailsDialog.java
/**
 * Copyright 2006-2011 Tuukka Haapasalo
 * 
 * This file is part of jAlkaMetri.
 * 
 * jAlkaMetri is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 * 
 * jAlkaMetri is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with jAlkaMetri (LICENSE.txt). If not, see <http://www.gnu.org/licenses/>.
 */
package fi.tuska.jalkametri.gui;

import java.text.DateFormat;

import android.app.Dialog;
import android.content.Context;
import android.content.res.Resources;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import fi.tuska.jalkametri.R;
import fi.tuska.jalkametri.data.Drink;
import fi.tuska.jalkametri.data.DrinkSelection;
import fi.tuska.jalkametri.data.DrinkSize;
import fi.tuska.jalkametri.util.StringUtil;
import fi.tuska.jalkametri.util.TimeUtil;

public class DrinkDetailsDialog extends Dialog {

    private DateFormat dateFormat;

    public DrinkDetailsDialog(Context context) {
        super(context);
        dateFormat = new TimeUtil(context).getDateFormatFull();
        setContentView(R.layout.show_event);

        Button okButton = (Button) findViewById(R.id.ok);
        okButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dismiss();
            }
        });
    }

    public void showDrinkSelection(DrinkSelection selection, boolean showTime) {
        Resources res = getContext().getResources();
        setTitle(selection.getIconText(res));
        
        Drink drink = selection.getDrink();
        // Drink name
        setDialogText(R.id.name, drink.getName());
        // Drink strength
        setDialogText(R.id.strength,
            String.format("%.1f %s", drink.getStrength(), res.getString(R.string.unit_percent)));

        DrinkSize size = selection.getSize();
        // Drink size name
        setDialogText(R.id.size_name, size.getName());
        // Drink size (volume, in liters)
        setDialogText(R.id.size, size.getFormattedSize(getContext()));

        // Portions
        setDialogText(
            R.id.portions,
            String.format("%.1f %s", selection.getPortions(getContext()),
                res.getString(R.string.unit_portions)));

        // Date
        {
            TextView drinkTime = (TextView) findViewById(R.id.date);
            if (showTime) {
                drinkTime.setVisibility(View.VISIBLE);
                setDialogText(R.id.date,
                    StringUtil.uppercaseFirstLetter(dateFormat.format(selection.getTime())));
            } else {
                drinkTime.setVisibility(View.GONE);
            }
        }

        // Icon
        {
            ImageView icon = (ImageView) findViewById(R.id.icon);
            int resID = DrinkIconUtils.getDrinkIconRes(selection.getIcon());
            icon.setImageResource(resID != 0 ? resID : R.drawable.launcher);
        }

        // Comment
        {
            TextView commentText = (TextView) findViewById(R.id.comment);
            String comment = drink.getComment();
            if ("".equals(comment))
                comment = null;
            if (comment != null) {
                commentText.setText(comment);
            }
            commentText.setVisibility(comment != null ? View.VISIBLE : View.GONE);
        }
    }

    public void setDialogText(int resID, String text) {
        TextView t = (TextView) findViewById(resID);
        if (t != null) {
            t.setText(text);
        }
    }

}

File: app/src/main/java/fi/tuska/jalkametri/gui/DrinkDetailsDialog.kt
package fi.tuska.jalkametri.gui

import android.app.DialogFragment
import android.content.res.Resources
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import fi.tuska.jalkametri.R
import fi.tuska.jalkametri.data.DrinkSelection
import fi.tuska.jalkametri.util.StringUtil
import fi.tuska.jalkametri.util.TimeUtil
import org.joda.time.format.DateTimeFormatter
import java.util.Locale

class DrinkDetailsDialog : DialogFragment() {

    private lateinit var model: ViewModel

    /*
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }
    */

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        val (selection, showTime) = readArgs(arguments)

        val v = inflater.inflate(R.layout.activity_show_event, container, false)
        model = ViewModel(selection, showTime, resources, v)

        val okButton = v.findViewById(R.id.ok) as Button
        okButton.setOnClickListener { dismiss() }
        return v
    }

    class ViewModel(val selection: DrinkSelection, val showTime: Boolean, val res: Resources, val view: View) {

        private val timeUtil = TimeUtil(res)
        private val dateFormat: DateTimeFormatter = timeUtil.dateFormatFull
        private val locale: Locale = timeUtil.locale

        init {

            val drink = selection.drink
            // Drink name
            setDialogText(R.id.name, drink.name)
            // Drink strength
            setDialogText(R.id.strength,
                    String.format(locale, "%.1f %s", drink.strength, res.getString(R.string.unit_percent)))

            val size = selection.size
            // Drink size name
            setDialogText(R.id.size_name, size.name)
            // Drink size (volume, in liters)
            setDialogText(R.id.size, size.getFormattedSize(res))

            // Portions
            setDialogText(
                    R.id.portions,
                    String.format(locale, "%.1f %s", selection.getPortions(view.context),
                            res.getString(R.string.unit_portions)))

            // Date
            run {
                val drinkTime = view.findViewById(R.id.date) as TextView
                if (showTime) {
                    drinkTime.visibility = View.VISIBLE
                    setDialogText(R.id.date,
                            StringUtil.uppercaseFirstLetter(dateFormat.print(selection.time)))
                } else {
                    drinkTime.visibility = View.GONE
                }
            }

            // Icon
            run {
                val icon = view.findViewById(R.id.icon) as ImageView
                val resID = DrinkIconUtils.getDrinkIconRes(selection.icon)
                icon.setImageResource(if (resID != 0) resID else R.mipmap.ic_launcher)
            }

            // Comment
            run {
                val commentText = view.findViewById(R.id.comment) as TextView
                var comment: String? = drink.comment
                if ("" == comment)
                    comment = null
                if (comment != null) {
                    commentText.text = comment
                }
                commentText.visibility = if (comment != null) View.VISIBLE else View.GONE
            }
        }

        private fun setDialogText(resID: Int, text: String) {
            val t = view.findViewById(resID) as TextView?
            if (t != null) {
                t.text = text
            }
        }
    }

    companion object {
        fun createDialog(selection: DrinkSelection, showTime: Boolean) = DrinkDetailsDialog().apply {
            arguments = writeArgs(Bundle(), selection, showTime)
        }

        fun writeArgs(bundle: Bundle, selection: DrinkSelection, showTime: Boolean) = bundle.apply {
            putSerializable("selection", selection)
            putBoolean("showTime", showTime)
        }

        fun readArgs(bundle: Bundle): Pair<DrinkSelection, Boolean> =
                Pair(bundle.getSerializable("selection") as DrinkSelection,
                        bundle.getBoolean("showTime"))
    }


}
